pilvytis/order_status_tracker.go

```diff
@@ -73,89 +55,145 @@ func NewStatusTracker(api orderProvider, identityProvider identityProvider, even
 		api:              api,
 		identityProvider: identityProvider,
 		eventBus:         eventBus,
-		orders:           []OrderSummary{},
+		orders:           make(map[string]map[uint64]OrderSummary),
+		failedSyncs:      make(map[identity.Identity]struct{}),
+		forceSync:        make(chan identity.Identity),
 		updateInterval:   updateInterval,
 		stopCh:           make(chan struct{}),
 	}
 }
 
-func (t *StatusTracker) getOrCreate(id uint64, identityAddress string) *OrderSummary {
-	for i := range t.orders {
-		if t.orders[i].ID == id {
-			return &t.orders[i]
-		}
+// SubscribeAsync subscribes to event to listen for unlock events.
+func (t *StatusTracker) SubscribeAsync(bus eventbus.Subscriber) error {
+	handleUnlockEvent := func(data identity.AppEventIdentityUnlock) {
+		t.UpdateOrdersFor(data.ID)
 	}
-	newOrder := OrderSummary{ID: id, IdentityAddress: identityAddress}
-	t.orders = append(t.orders, newOrder)
-	return &t.orders[len(t.orders)-1]
+
+	// Handle the unlock event so that we load any orders for identities that just launched without
+	// waiting for the thread to re-execute.
+	return bus.SubscribeAsync(identity.AppTopicIdentityUnlock, handleUnlockEvent)
 }
 
-// Track order status updates.
+// Track will block and start tracking orders.
 func (t *StatusTracker) Track() {
-	t.lock.Lock()
-	defer t.lock.Unlock()
-	if t.tracking {
-		return
-	}
-	t.tracking = true
-	go func() {
-		log.Info().Msg("Tracking order statuses...")
-		updateTicker := time.NewTicker(t.updateInterval)
-		defer updateTicker.Stop()
-		for {
-			select {
-			case <-t.stopCh:
-				return
-			case <-updateTicker.C:
-				err := t.update()
-				if err != nil {
-					log.Err(err).Msg("Could not update order statuses")
+	for {
+		select {
+		case <-t.stopCh:
+			return
+		case id := <-t.forceSync:
+			t.refreshAndUpdate(id)
+		case <-time.After(t.updateInterval):
+			for _, id := range t.identityProvider.GetIdentities() {
+				if !t.identityProvider.IsUnlocked(id.Address) {
+					continue
+				}
+				if !t.needsRefresh(id) {
+					continue
 				}
+
+				t.refreshAndUpdate(id)
 			}
 		}
-	}()
+	}
+}
+
+// UpdateOrdersFor sends a notification to the main running thread to
+// sync orders for the given identity.
+func (t *StatusTracker) UpdateOrdersFor(id identity.Identity) {
+	t.forceSync <- id
+}
+
+// Stop stops the status tracker
+func (t *StatusTracker) Stop() {
+	close(t.stopCh)
```

Comment:
do a once here, we're notorious for stopping everything twice.

Reply:
thats very fair